fs: allow int64 offset in fs.read/readSync/write/writeSync

zbjornson opened the pull request:
Since v10.10.0, 'buf' can be any DataView, meaning the largest byteLength can be Float64Array.BYTES_PER_ELEMENT * kMaxLength = 17,179,869,176.

Fixes #26563

~~Eventually, the same change applies to fs.write*. If this current change is acceptable, I'll make the corresponding change there.~~ Done

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included - No. I can add a test but I don't see a way to do it without requiring a large amount of disk space and memory. benchmark N/A.
- [ ] documentation is changed or added - No. If this is significant enough to note in the history section, let me know.
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented on lib/fs.js:
```diff
@@ -454,8 +454,8 @@ function read(fd, buffer, offset, length, position, callback) {
-  offset |= 0;
-  length |= 0;
+  offset = Math.round(offset);
+  length = Math.round(length);
```
`|= 0` was likely originally used to handle undefined values. This would now be changed to `NaN` instead of `0`.

I would also rather validate that we actually received an integer value instead of rounding.

I suggest to change this to:

```js
if (offset == null) {
  offset = 0;
} else {
  validateInteger(offset, 'offset');
}
```

bnoordhuis commented on src/node_file.cc:
```diff
@@ -1846,12 +1846,12 @@ static void Read(const FunctionCallbackInfo<Value>& args) {
 
-  CHECK(args[2]->IsInt32());
-  const size_t off = static_cast<size_t>(args[2].As<Int32>()->Value());
+  CHECK(args[2]->IsNumber());
```
Saying it's an int64 and using `IsNumber()` is wrong-ish because the latter signals it's a double and those are only precise up to 2**53-1.

I wouldn't be surprised if we conflate them in other places too. Perhaps it's time for V8 to grow an `IsInt64()` method or for us to grow a utility method that checks the same.

zbjornson replied:
It's already checked in the JS wrapper; is that not enough?

(I copied what's done for `position` a few lines down.)

bnoordhuis replied:
> It's already checked in the JS wrapper; is that not enough?

It's fragile and non-obvious, IMO.

Aside: maybe `validateInteger()` should be renamed to `validateSafeInteger()` - just 'integer' doesn't convey much except that it rejects rationals.

> It's also tested on the next line to be less than the buffer length (guaranteed to be safe int).

That's another buglet: `size_t` might be 32 bits. Narrowing from `int64_t` to `size_t` can be lossy.

zbjornson replied:
Is this close to what you have in mind?
```cpp
void CheckSafePositiveJsInt(v8::Value v) {
  CHECK(v->IsNumber());
  double v_d = v.As<Number>()->Value();
  CHECK_EQ(std::trunc(v_d), v_d); // is int
  int64_t v_i64 = static_cast<int64_t>(v_d);
  CHECK_GE(v_i64, 0); // is positive int
  CHECK_LE(v_i64, MAX_SAFE_JS_INTEGER); // is positive safe JS int
}

// in Read()
CheckSafePositiveJsInt(args[2]);
// now safe to cast to uint64_t (likely equivalent of size_t)
uint64_t off_u64 = static_cast<uint64_t>(args[2].As<Integer>()->Value());
CHECK_LT(off_u64, buffer_length); // buffer_length promoted on 32-bit arch
// buffer_length is size_t, so now safe to cast to size_t
size_t off = static_cast<size_t>(off_u64);

```

bnoordhuis replied:
Yes, that looks good.

ryzokuken commented on src/node_file.cc:
```diff
@@ -1846,12 +1847,14 @@ static void Read(const FunctionCallbackInfo<Value>& args) {
-  CHECK(args[3]->IsInt32());
-  const size_t len = static_cast<size_t>(args[3].As<Int32>()->Value());
+  CHECK(args[3]->IsUint32());
+  const size_t len = static_cast<size_t>(args[3].As<Uint32>()->Value());
```
💯

zbjornson commented:
Read for review again. I ended up changing directions a bit: `offset` can be 2<sup>53</sup>-1, but `len` can only be `uint32` because there's no point to supporting the full range:

1. `uv_buf_init` only takes `unsigned int` length.
1. On UNIX-y platforms, `uv_buf_init` can be simply changed to take `size_t` (`uv_buf_t` already has `size_t` *`len`*), but Linux `read()`/`pread()` will only transfer up to 2 GB - 4 KB per call.
1. On Windows, `ReadFile` can only transfer up to 4 GB.

User-space tiling with `fs.read` is easy and, with this patch allowing big `offset` values, possible; so I think this is a reasonable outcome.

(When I have time I can work on separate PRs for `write` and streams.)

ryzokuken commented on src/util-inl.h:
```diff
@@ -491,6 +492,15 @@ void ArrayBufferViewContents<T, S>::Read(v8::Local<v8::ArrayBufferView> abv) {
 }
 
+inline void CheckSafePositiveJsInt(v8::Local<v8::Value> v) {
+  CHECK(v->IsNumber());
```
Should this function return a `v8::Maybe` instead of just crashing the code? Or maybe return a `bool`?

I'm just looking at it, and it's not very obvious that this will crash node.
Furthermore, if it's not a safe positive integer, I guess we should just inform the caller and they could appropriately handle the situation, well... situationally.

Just seems to be a more generalized approach to me, but I'm happy as long as it work anyway.

zbjornson replied:
Good idea; changed.

zbjornson commented:
Changed implementation a little bit so the `IsSafeJsInt` function mirrors ECMA262's `Number.isSafeInteger` spec.

zbjornson commented:
Revised again to just widen the `offset` parameter, but leave `length` alone. Widening `length` needs a lot of changes because several functions in libuv and node_file.cc have `int` return types that overflow. Since the change would only benefit Windows but not Linux (idk about MacOS), I don't think it's worth the change.

BridgeAR commented:
@zbjornson this requires a rebase due to recent changes. It would also be great if you could add a test that verifies that `NaN` will throw (it used to be coerced to 0).

zbjornson commented:
@BridgeAR would you prefer that, or going back to coercing `NaN` to 0 to minimize changes in behavior?

BridgeAR commented:
@zbjornson handling `NaN` as `0` seems wrong and it's almost certainly a user error in some calculation. Throwing in such situations seems correct to me (So yes, I prefer that). It's just a breaking change.

zbjornson commented:
I've made the 2nd commit coerce NaN to 0 to preserve behavior, and made the semver-major change in a 3rd commit. I was hoping the 1st and 2nd commits (int64 offset) can land in v10.

BridgeAR commented:
@zbjornson sorry, this needs another rebase. The first commit could definitely be backported but we normally do not backport any part of a semver-major PR (only if necessary). The second commit would still be a breaking change though as it also invalides `''`, `false`, 'true', `Infinity`, `'0'`, '100' and any other number string.

I personally would be fine to accept it as patch in case strings get excluded but the string conversion is something more likely (I do not worry about NaN all that much). I can not speak for anyone else though @addaleax @bnoordhuis @ryzokuken @nodejs/tsc.

zbjornson commented:
No problem, rebased.

I *think* I got the 2nd commit to be non-breaking finally. If it's too much hassle/can't land in 10.x anyway, the 2nd and 3rd commits can be squashed.

I have the equivalent `offset` changes prepared for fs.write & friends. I was planning to put those in a separate PR, but let me know if you'd rather them be in this PR.

Finally, since this PR throws for non-int/null/undefined `offset`s, it seems reasonable to do the same for `length`. I don't want to creep on scope too much. Let me know your preference.

BridgeAR commented on test/parallel/test-fs-util-validateoffsetlengthwrite.js:
```diff
@@ -22,22 +22,6 @@ const { kMaxLength } = require('buffer');
-               `It must be <= ${kMaxLength - offset}. Received ${length}`
-    }
-  );
-}
```
Does this test have to be removed? To me it looks like it should still pass but maybe with a slightly different message?

zbjornson replied:
I think it's redundant with the test below it, right? This one tested the case of `byteLength > kMaxLength`, which is no longer a requirement.

BridgeAR commented on lib/fs.js:
```diff
@@ -537,8 +547,11 @@ function write(fd, buffer, offset, length, position, callback) {
     callback = maybeCallback(callback || position || length || offset);
-    if (typeof offset !== 'number')
+    if (typeof offset !== 'number') {
       offset = 0;
```
Since this PR is still semver-major I suggest to change this check (and the others that work the identical to this one) to:

```js
if (offset == null) {
  offset = 0;
} else {
  ...
}
```

zbjornson commented:
Changed to `if (offset == null) offset = 0` and rebased. Thanks!

zbjornson commented:
@BridgeAR is there anything else needed for this to land?

jasnell commented on src/util.h:
```diff
@@ -168,6 +168,10 @@ void DumpBacktrace(FILE* fp);
 
 #define UNREACHABLE() ABORT()
 
+constexpr int64_t kMaxSafeJsInteger = 9007199254740991;  // 2^53-1
```
would likely be worthwhile to add a comment about where this value comes from

thefourtheye commented on src/util-inl.h:
```diff
@@ -502,6 +503,17 @@ void ArrayBufferViewContents<T, S>::Read(v8::Local<v8::ArrayBufferView> abv) {
+  if (std::isnan(v_d)) return false;
+  if (std::isinf(v_d)) return false;
+  if (std::trunc(v_d) != v_d) return false;  // not int
+  if (std::abs(v_d) <= static_cast<double>(kMaxSafeJsInteger)) return true;
```
Nit: Simply returning the result of this condition also should work.

zbjornson replied:
Very true. This method is a verbatim port of the referenced part of ECMA-262 (https://www.ecma-international.org/ecma-262/9.0/index.html#sec-number.issafeinteger) though; I thought that would be clearer to follow exactly.

zbjornson commented on lib/fs.js:
```diff
@@ -633,7 +649,7 @@ function truncate(path, len, callback) {
     len = 0;
   }
 
-  validateInteger(len, 'len');
```
bnoordhuis suggested the change (https://github.com/nodejs/node/pull/26572#discussion_r264536405), and I isolated them in a separate commit (intending all commits in this PR to be merged, not squashed).

zbjornson commented on src/node_file.cc:
```diff
@@ -1649,9 +1649,11 @@ static void WriteBuffer(const FunctionCallbackInfo<Value>& args) {
+  CHECK(IsSafeJsInt(args[2]));
+  const int64_t off_64 = args[2].As<Integer>()->Value();
+  CHECK_GE(off_64, 0);
+  CHECK_LE(static_cast<uint64_t>(off_64), buffer_length);
```
`size_t` might only be a `uint32` (32-bit arch). Once we know `off_64 < [size_t]buffer_length`, then it's safe to convert it to a `size_t`.

zbjornson commented:
* Fixed the test failure. @BridgeAR you might want to look at the offset check (line 550R of lib/fs.js, in `write`), I had to tweak it to support all of the allowed overloads.
* Rebased on master.
* Added a comment explaining kMaxSafeJsInteger per @jasnell.

zbjornson commented:
@BridgeAR sorry to pester but worried this PR has fallen into purgatory. Anything else needed for this to land?

ryzokuken commented:
@zbjornson since nobody really blocked this PR, I think it could be landed. That said, we'd need to run CI again, I'm afraid.

zbjornson commented:
I've rebased this on master to address conflicts. Could someone start CI please?

zbjornson commented:
Thanks. Could someone start CI one more time please? I fixed the real failure (mistake in the test), and I think the other two (http2) are flaky tests looking at other recent runs.

Trott commented:
CITGM for master: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1945/
CITGM for this PR (queued--will 404 until a worker is available): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1946/

Trott commented:
CITGM for this PR is failing in a weird way unless there's a typo in my refspec I'm not seeing...

/ping @nodejs/build

```console
22:57:55 Building remotely on test-packetnet-ubuntu1604-x64-1 (Ubuntu Ubuntu-16.04 amd64-Ubuntu jenkins-workspace amd64-Ubuntu-16.04 16.04 amd64) in workspace /home/iojs/build/workspace/citgm-smoker
22:57:55 using credential 96d5f81c-e9ad-45f7-ba5d-bc8107c0ae2c
22:57:55 FATAL: Invalid refspec refs/pull//26572/head
22:57:55 java.lang.IllegalArgumentException: Invalid refspec refs/pull//26572/head
22:57:55 	at org.eclipse.jgit.transport.RefSpec.checkValid(RefSpec.java:539)
22:57:55 	at org.eclipse.jgit.transport.RefSpec.<init>(RefSpec.java:191)
22:57:55 	at org.eclipse.jgit.transport.RefSpec.<init>(RefSpec.java:226)
22:57:55 	at hudson.plugins.git.GitSCM.getRefSpecs(GitSCM.java:592)
22:57:55 	at hudson.plugins.git.GitSCM.getParamExpandedRepo(GitSCM.java:510)
22:57:55 	at hudson.plugins.git.GitSCM.getParamExpandedRepos(GitSCM.java:497)
22:57:55 	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1117)
22:57:55 	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1175)
22:57:55 	at hudson.scm.SCM.checkout(SCM.java:504)
22:57:55 	at hudson.model.AbstractProject.checkout(AbstractProject.java:1208)
22:57:55 	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:574)
22:57:55 	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
22:57:55 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:499)
22:57:55 	at hudson.model.Run.execute(Run.java:1818)
22:57:55 	at hudson.matrix.MatrixBuild.run(MatrixBuild.java:323)
22:57:55 	at hudson.model.ResourceController.execute(ResourceController.java:97)
22:57:55 	at hudson.model.Executor.run(Executor.java:429)
22:57:55 Collecting metadata...
22:57:55 Metadata collection done.
22:57:55 Notifying upstream projects of job completion
22:57:55 Finished: FAILURE
```

Trott commented:
Never mind! I see now! Double-slash! I should wear my reading glasses when doing this stuff, apparently.

CITGM for this PR: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1948/

Trott commented:
Landed in c3b8e50...5e3b4d6. 🎉
